pkg/controller/globalrole/globalrole_controller.go

```diff
@@ -0,0 +1,361 @@
+		return err
+	}
+
+	var federatedGlobalRole iamv1alpha2.FederatedRole
+
+	if err = runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(*unstructured.Unstructured).Object, &federatedGlobalRole); err != nil {
+		klog.Error(err)
+		return err
+	}
+
+	if !reflect.DeepEqual(federatedGlobalRole.Spec.Template.Rules, globalRole.Rules) ||
+		!reflect.DeepEqual(federatedGlobalRole.Spec.Template.Labels, globalRole.Labels) ||
+		!reflect.DeepEqual(federatedGlobalRole.Spec.Template.Annotations, globalRole.Annotations) {
+
+		federatedGlobalRole.Spec.Template.Rules = globalRole.Rules
+		federatedGlobalRole.Spec.Template.Annotations = globalRole.Annotations
+		federatedGlobalRole.Spec.Template.Labels = globalRole.Labels
+
+		return c.updateFederatedGlobalRole(&federatedGlobalRole)
+	}
+
+	return nil
+}
+
+func (c *Controller) createFederatedGlobalRole(globalRole *iamv1alpha2.GlobalRole) error {
+	federatedGlobalRole := &iamv1alpha2.FederatedRole{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       iamv1alpha2.FedGlobalRoleKind,
+			APIVersion: iamv1alpha2.FedGlobalRoleResource.Group + "/" + iamv1alpha2.FedGlobalRoleResource.Version,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: globalRole.Name,
+		},
+		Spec: iamv1alpha2.FederatedRoleSpec{
+			Template: iamv1alpha2.RoleTemplate{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      globalRole.Labels,
+					Annotations: globalRole.Annotations,
+				},
+				Rules: globalRole.Rules,
+			},
+			Placement: iamv1alpha2.Placement{
+				ClusterSelector: iamv1alpha2.ClusterSelector{},
+			},
+		},
+	}
+
+	err := controllerutil.SetControllerReference(globalRole, federatedGlobalRole, scheme.Scheme)
+
+	if err != nil {
+		return err
+	}
+
+	data, err := json.Marshal(federatedGlobalRole)
+
+	if err != nil {
+		return err
+	}
+
+	cli := c.k8sClient.(*kubernetes.Clientset)
+	err = cli.RESTClient().Post().
+		AbsPath(fmt.Sprintf("/apis/%s/%s/%s", iamv1alpha2.FedGlobalRoleResource.Group,
+			iamv1alpha2.FedGlobalRoleResource.Version, iamv1alpha2.FedGlobalRoleResource.Name)).
+		Body(data).
+		Do().Error()
+
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			return nil
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (c *Controller) updateFederatedGlobalRole(federatedGlobalRole *iamv1alpha2.FederatedRole) error {
+
+	data, err := json.Marshal(federatedGlobalRole)
+
+	if err != nil {
+		return err
+	}
+
+	cli := c.k8sClient.(*kubernetes.Clientset)
+
+	err = cli.RESTClient().Put().
+		AbsPath(fmt.Sprintf("/apis/%s/%s/%s/%s", iamv1alpha2.FedGlobalRoleResource.Group,
+			iamv1alpha2.FedGlobalRoleResource.Version, iamv1alpha2.FedGlobalRoleResource.Name,
+			federatedGlobalRole.Name)).
+		Body(data).
+		Do().Error()
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (c *Controller) ensureNotControlledByKubefed(globalRole *iamv1alpha2.GlobalRole) error {
+	if globalRole.Labels[constants.KubefedManagedLabel] != "false" {
+		if globalRole.Labels == nil {
```

Comment:
if `globalRole.Labels == nil`, the statement above will throw a nil pointer exception.

Reply:
I'm sure there won't be a nil pointer here.